File: backend-app/deployment-service/src/main/java/nl/saxion/disaster/deploymentservice/model/DeploymentRequest.java
package nl.saxion.disaster.deploymentservice.model;

import jakarta.persistence.*;
import lombok.Data;
import nl.saxion.disaster.deploymentservice.enums.DeploymentRequestStatus;
import nl.saxion.disaster.deploymentservice.enums.IncidentSeverity;
import nl.saxion.disaster.deploymentservice.enums.ResponseUnitType;

import java.util.Date;

@Entity
@Table(name = "deployment_requests")
@Data
public class DeploymentRequest {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long requestId;

    @Column(nullable = false)
    private Long incidentId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "deployment_order_id", nullable = false)
    private DeploymentOrder deploymentOrder;

    @Column(nullable = false)
    private Long requestedBy;

    @Column(nullable = false)
    @Temporal(TemporalType.TIMESTAMP)
    private Date requestedAt;

    @Column(nullable = false)
    private Long targetDepartmentId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private IncidentSeverity priority;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ResponseUnitType requestedUnitType;

    @Column(nullable = false)
    private int requestedQuantity;

    private Long assignedUnitId;
    private Long assignedBy;
    private Date assignedAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DeploymentRequestStatus status;

    private String notes;
}

File: client-app/src/api/deploymentRequest.ts
import { BaseApi } from '@/api/base';
import type { DeploymentRequest } from '@/types/deployment';
import { DeploymentRequestStatus } from '@/types/deployment';
import { ResponseUnitType } from '@/types/responseUnit';
import { IncidentSeverity } from '@/types/incident';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL;

// --- DTOs ---
interface DeploymentRequestDTO {
	requestId: number;
	incidentId: number;
	deploymentOrderId: number;
	requestedBy: number;
	requestedAt: string;
	targetDepartmentId: number;
	priority: IncidentSeverity;
	requestedUnitType: ResponseUnitType;
	requestedQuantity: number;
	assignedUnitId?: number;
	assignedBy?: number;
	assignedAt?: string;
	status: DeploymentRequestStatus;
	notes?: string;
}

// --- Mapping ---
function mapDeploymentRequestDTO(dto: DeploymentRequestDTO): DeploymentRequest {
	return {
		requestId: dto.requestId,
		incidentId: dto.incidentId,
		deploymentOrderId: dto.deploymentOrderId,
		requestedBy: dto.requestedBy,
		requestedAt: new Date(dto.requestedAt),
		targetDepartmentId: dto.targetDepartmentId,
		priority: dto.priority,
		requestedUnitType: dto.requestedUnitType,
		requestedQuantity: dto.requestedQuantity,
		...(dto.assignedUnitId !== undefined ? { assignedUnitId: dto.assignedUnitId } : {}),
		...(dto.assignedBy !== undefined ? { assignedBy: dto.assignedBy } : {}),
		...(dto.assignedAt ? { assignedAt: new Date(dto.assignedAt) } : {}),
		status: dto.status,
	};
}

// --- API ---
const deploymentRequestApi = new BaseApi(`${API_BASE_URL}/deployment-requests`);

export async function getDeploymentRequestById(id: number): Promise<DeploymentRequest> {
	const dto = await deploymentRequestApi.get<DeploymentRequestDTO>(`/${id}`);
	return mapDeploymentRequestDTO(dto);
}
